snapcraft/commands/refresh-expiring-certs

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env bash
```

Comment:
I deliberated a lot about whether to implement this in `microovn` itself or as a service script and I decided to go with the script for following reasons:
* I **really** wanted to leverage existing functionality of `openssl` that just tells you if the certificate expires within specified time period or not. This saves us a lot of implementation details that we'd have to deal with if we were to implement this check ourselves.
* This is a functionality that is not useful to user in day-to-day operations. We'd be adding CLI command that would not be useful to end user, unnecessarily cluttering the command line UI.

I realize that these are not the strongest points but it's what tipped me towards the script. To your points

> without any meaningful way of testing

I'd argue that we can test this quite easily by passing argument `-t 4000` (~11 years) which would guarantee that both CA and the service certificates would be considered expiring and we could observe whether the automatic refresh process works.

>  it appears that the previous commit of adding JSON output format is also done only to provide support for the processing in this bash script.

You are right, that was the main motivation, although I'd say that having machine-readable output is never bad thing and can come in handy later even if there's no other clear use-case for it right now.

>  the addition of the jq stage-package

The `jq` package itself is very minimalistic. It has only one dependency (outside of libc) and both main package and its dependencies are just couple hundred kB in size combined. 

> The CA regenration implementation could for example also have an extra argument only allowing the current microcluster leader do the CA regeneration, avoiding race conditions.

I tried to mitigate the possible race condition by specifying time range for the daily check so that not all members execute it at the same time (we can expand the time range to further lower the race condition probability). Adding the self-imposed client-side limitation doesn't sound very good to me. It'd be again something that's not very useful to regular user and kind-of forced into the CLI for this one specific task.
I didn't test this but I suspect that the race-condition might not even be that sever. The "regenerate-ca" process has 3 steps:
* Issue new CA (distributed via dqlite)
* request peers to issue new service certificates
* issue local service certificates
Now if in the middle of this process comes another member and executes the same, my suspicion is that it'll cause a bit of thrashing but everyone will eventually end up with the CA and certificates that were generated by the member that initiated the process later. 

All this being said I'm not opposed to implementing it in `microovn` itself, as I mentioned I considered it heavily. But as you said, It'd probably push the implementation out of this pulse. Let me know if you still think that the service script is not a good idea.

Reply:
TLDR; I'm still a bit skeptic to the bash approach, but it does serve the purpose now, and we can always change it later, so will not block on it now.

> I deliberated a lot about whether to implement this in `microovn` itself or as a service script and I decided to go with the script for following reasons:
> 
>     * I **really** wanted to leverage existing functionality of `openssl` that just tells you if the certificate expires within specified time period or not. This saves us a lot of implementation details that we'd have to deal with if we were to implement this check ourselves.

The go code could still execute the `openssl` binary to do the work. My main motivation for wanting it there is just to avoid the need for exporting all this data and process it in bash.

>     * This is a functionality that is not useful to user in day-to-day operations. We'd be adding CLI command that would not be useful to end user, unnecessarily cluttering the command line UI.

That is a fair point, I do think a `--reissue-only-if-the-certificate-is-about-to-expire` (obviously has to have a different name ;)) option could be useful for the end user though.

> > without any meaningful way of testing
> 
> I'd argue that we can test this quite easily by passing argument `-t 4000` (~11 years) which would guarantee that both CA and the service certificates would be considered expiring and we could observe whether the automatic refresh process works.

If you say so, I have only had regrets when it comes to bash and testing, but I guess it's not impossible.

> > it appears that the previous commit of adding JSON output format is also done only to provide support for the processing in this bash script.
> 
> You are right, that was the main motivation, although I'd say that having machine-readable output is never bad thing and can come in handy later even if there's no other clear use-case for it right now.

Machine readable stuff is indeed good, but you're also creating an API that an end user may rely on. What if we at some point in the future do not have this information available anymore (for example because doing cert lifecycle somewhere else), would we break anyone's workflow?

> > the addition of the jq stage-package
> 
> The `jq` package itself is very minimalistic. It has only one dependency (outside of libc) and both main package and its dependencies are just couple hundred kB in size combined.

:+1:

> > The CA regenration implementation could for example also have an extra argument only allowing the current microcluster leader do the CA regeneration, avoiding race conditions.
> 
> I tried to mitigate the possible race condition by specifying time range for the daily check so that not all members execute it at the same time (we can expand the time range to further lower the race condition probability). Adding the self-imposed client-side limitation doesn't sound very good to me. It'd be again something that's not very useful to regular user and kind-of forced into the CLI for this one specific task. I didn't test this but I suspect that the race-condition might not even be that sever. The "regenerate-ca" process has 3 steps:
> 
>     * Issue new CA (distributed via dqlite)
> 
>     * request peers to issue new service certificates
> 
>     * issue local service certificates
>       Now if in the middle of this process comes another member and executes the same, my suspicion is that it'll cause a bit of thrashing but everyone will eventually end up with the CA and certificates that were generated by the member that initiated the process later.

Synchronization in clustered systems is hard, so adding CLI options to enable users to deal with it is not a bad thing imo. there is precedent for `--leader-only` / `--no-leader-only` type options in the upstream project as well, so it would be familiar. The end user would have the same problem if they wanted to manage the lifecycle themselves.

> All this being said I'm not opposed to implementing it in `microovn` itself, as I mentioned I considered it heavily. But as you said, It'd probably push the implementation out of this pulse. Let me know if you still think that the service script is not a good idea.

Yep, let's keep your current implementation and revisit in the future if needed.